File: app/src/main/java/com/bijoysingh/quicknote/items/RecyclerItem.java
package com.bijoysingh.quicknote.items;

import com.bijoysingh.quicknote.R;
import com.bijoysingh.quicknote.recyclerview.EmptyRecyclerHolder;
import com.bijoysingh.quicknote.recyclerview.FileImportViewHolder;
import com.bijoysingh.quicknote.recyclerview.NoteRecyclerHolder;
import com.github.bijoysingh.starter.recyclerview.MultiRecyclerViewControllerItem;

import java.util.ArrayList;
import java.util.List;

public abstract class RecyclerItem {
  public enum Type {
    NOTE,
    EMPTY,
    FILE,
  }

  abstract public Type getType();

  public static List<MultiRecyclerViewControllerItem<RecyclerItem>> getList(boolean staggered) {
    List<MultiRecyclerViewControllerItem<RecyclerItem>> list = new ArrayList<>();
    list.add(new MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
        .viewType(Type.NOTE.ordinal())
        .layoutFile(staggered ? R.layout.item_note_staggered : R.layout.item_note)
        .holderClass(NoteRecyclerHolder.class)
        .build());
    list.add(new MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
        .viewType(Type.EMPTY.ordinal())
        .layoutFile(R.layout.item_no_notes)
        .holderClass(EmptyRecyclerHolder.class)
        .spanSize(2)
        .build());
    list.add(new MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
        .viewType(Type.FILE.ordinal())
        .layoutFile(R.layout.item_import_file)
        .holderClass(FileImportViewHolder.class)
        .build());
    return list;
  }
}

File: app/src/main/java/com/bijoysingh/quicknote/items/RecyclerItem.kt
package com.bijoysingh.quicknote.items

import com.bijoysingh.quicknote.R
import com.bijoysingh.quicknote.recyclerview.*
import com.github.bijoysingh.starter.recyclerview.MultiRecyclerViewControllerItem
import java.util.*

abstract class RecyclerItem {

  abstract val type: Type

  enum class Type {
    NOTE,
    EMPTY,
    FILE,
    INFORMATION
  }

  companion object {
    fun getList(
        staggered: Boolean,
        isTablet: Boolean): List<MultiRecyclerViewControllerItem<RecyclerItem>> {
      val list = ArrayList<MultiRecyclerViewControllerItem<RecyclerItem>>()
      list.add(MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
          .viewType(Type.NOTE.ordinal)
          .layoutFile(if (staggered && !isTablet) R.layout.item_note_staggered else R.layout.item_note)
          .holderClass(NoteRecyclerHolder::class.java)
          .build())
      list.add(MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
          .viewType(Type.EMPTY.ordinal)
          .layoutFile(R.layout.item_no_notes)
          .holderClass(EmptyRecyclerHolder::class.java)
          .spanSize(2)
          .build())
      list.add(MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
          .viewType(Type.INFORMATION.ordinal)
          .layoutFile(R.layout.item_information)
          .holderClass(InformationRecyclerHolder::class.java)
          .spanSize(2)
          .build())
      list.add(MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
          .viewType(Type.FILE.ordinal)
          .layoutFile(R.layout.item_import_file)
          .holderClass(FileImportViewHolder::class.java)
          .build())
      return list
    }

    fun getSelectableList(
        staggered: Boolean,
        isTablet: Boolean): List<MultiRecyclerViewControllerItem<RecyclerItem>> {
      val list = ArrayList<MultiRecyclerViewControllerItem<RecyclerItem>>()
      list.add(MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
          .viewType(Type.NOTE.ordinal)
          .layoutFile(if (staggered && !isTablet) R.layout.item_note_staggered else R.layout.item_note)
          .holderClass(SelectableNoteRecyclerViewHolder::class.java)
          .build())
      list.add(MultiRecyclerViewControllerItem.Builder<RecyclerItem>()
          .viewType(Type.EMPTY.ordinal)
          .layoutFile(R.layout.item_no_notes)
          .holderClass(EmptyRecyclerHolder::class.java)
          .spanSize(2)
          .build())
      return list
    }
  }
}
